### PR TITLE
PDEs in arbitrary dimensions and vectorized differentiation for optimization

### PR DESCRIPTION
**Pull request by znicolaou**
I implemented several adjustments to enhance the performance of the PDELibrary in higher dimensions.

1. I added an axis option to the FiniteDifference class, with default value axis=0. This code is essentially the same as the current version, but new array slice tuples were introduced to allow the differentiation axis to be set. The previous version assumed the input was 2 dimensional (time along the zero axis and feature along the second axis), which was based on compatibility with sklearn. However, compatibility with sklearn is already enforced in feature libraries, since the transform function reshapes the features into a 2-dimensional matrix. The advantage of this is that the spatial derivatives can now be calculated in a completely vectorized fashion, eliminating the need for slow python for loops in the PDELibrary function (see 3). 

2. I added a SpectralDerivative class to the differentiation module, which performs a vectorized spectral derivative of a given axis. This is equivalent to a maximal-order finite-difference derivative, but can be performed in O(NlogN) time rather than O(N^2) by leveraging FFTs. This is most useful for spatial derivatives in periodic domains, where it can significantly reduce the number of mesh points required to achieve accuracy and speed up numerical integration.

3. I wrote a new PDELibrary class which can be applied to any number of spatial dimensions, using the old class as a template. This just involves some fancy array reshaping and slicing. In addition to adding functionality, it significantly shortens and simplifies the code. Using the new vectorized derivatives, I found that the PDELibrary.transform function can run up to 50x faster for a case with 2 spatial dimensions. I have not yet implemented the weak for arbitrary dimensions. 

4. I copied the the current version of PDELibrary to a new WeakPDELibrary class, in order to maintain its function. I am going to see if it will be possible to rewrite this class for arbitrary dimensions before I make a final pull request.

5. I implemented a small improvement to the TensoredLibrary by vectorizing the calculation of outer products using numpy slicing.

6. I verified on the example notebooks 1 and 10 that the results with the new code are unchanged.

### CODE REVIEW

**Comment by akaptano** [comment]
Awesome job with this! I think first thing to do is add unit tests to the tests/ directory for the functionality you have introduced so far, verifying both that the new functionality reproduces the old, and the speedup is as advertised. I'll take a look at this next week but feel free to add new commits if you work on it before then. And let me know if you have questions about adding unit tests :)

**Comment by akaptano** [comment]
I will also point out that the eventual goal here should be to speed up the weak form stuff since this is 1) the most useful technique for PDEs and noisy data and 2) it is currently quite slow in 2 or 3 spatial dimensions. Exciting!

**Comment by akaptano** [comment]
Two other smaller notes:

Your adjustment (2) may actually be a duplicate with existing functionality, since the custom differentiation methods allows users to use differentiators from the python "derivative" package... see the Example 5 Jupyter notebook for an example of using spectral differentiation with this package. Unless there is some added functionality with your new implementation, we might consider chopping this. 

One note on changing PDELibrary or writing a new weak form library... we also want to maintain the weak form functionality for systems of ODEs, since it still gives big improvements when noise is present. We might consider a new WeakForm Library that can take combinations of different PDE or ODE libraries and weak-form each... but this is sort of what the GeneralizedLibrary already facilitates. The downside of keeping it as "PDELibrary" or "WeakPDELibrary" is that we seem to be implying the weak form works only for PDEs.

**Comment by znicolaou** [comment]
Thanks, Alan! Will check out the how the unit testing stuff works hopefully soon too. I think the weak form may be some effort to figure out a general procedure for arbitrary dimension, but will definitely try to crack it. Also, makes a lot of sense to me to have some sort of weak form library for both ODEs and PDEs. Still a work in progress...

**Comment by akaptano** [comment]
Hey Zach, I'm going to make this fork into a new branch on the main dynamicslab GitHub, since this makes it easier to switch /merge/etc. between the main branch and this one. If that is alright, let's use this new branch going forward.

**Comment by znicolaou** [comment]
I don't think I have write access to the dynamicslab repository yet, but feel free to go ahead with a branch there! I'll reach out soon about getting access.

**Comment by akaptano** [comment]
Ah yes, sorry you need to reach out to Brian. Updating this pull request to reflect the branch on the main repo.

**Comment by znicolaou** [comment]
Haven't linted before, but will start now :) thanks!

**Comment by znicolaou** [comment]
I managed to get all the unit test to run without errors. Since I separated to PDELibrary and WeakPDELibrary, this required editing the tests some as well. I'll resolve the linting errors next soon make a push soon!

**Comment by akaptano** [comment]
Thanks! Sorry for the slowness on my end. After you make push your linting fixes, let me know and I'll go through and further polish + test things.

**Comment by znicolaou** [comment]
I reached out about getting write access here, but still waiting for now. In the meanwhile, I pushed to my fork on the pde_optimization branch of https://github.com/znicolaou/pysindy, and I think you should be able to merge from there again when you want. No rush to look! It's break time after all ;)

**Comment by znicolaou** [comment]
Finally, no more linting errors now. The pytest runs successfully on my machine, but I get some failures on test_trapping_sr3_quadratic_library in test_optimizers.py on github now.  I don't think that test involves any of the files I changed. Any suggestions?

Also, I did get write access now, so I can merge here in the future--thanks!

**Comment by akaptano** [comment]
Awesome job. Depends on the error you are getting but your computer might be unhappy with some CVXPY subroutines. So is the pde_optimization on the main PySINDy branch up to date? If not, please merge your most recent changes, because I am adding a big set of changes tomorrow. Went through the weak form stuff and otherwise did some clean up.

**Comment by znicolaou** [comment]
My current version is on the znicolaou/pysindy pde_optimization branch, which is synced with this pull request and runs the test here when I push there. I'll take a minute to merge with the pde_optimization branch in the dynamicslab repository tomorrow.  

My last message was confusing about the pytest errors, but on my local computer, there are no errors in any of the test (including test_optimizers.py). But there are some kind of convergence errors when GitHub runs the test online here: https://github.com/dynamicslab/pysindy/runs/4578380301.

**Comment by znicolaou** [comment]
I changed the base on this pull request to the dynamicslab/pde_optimization branch and merged the fork, so this pull request was closed. Alan, feel free to merge with your changes and open a new pull request for review.

**Comment by akaptano** [comment]
Sounds good... there is also a "revert" pde optimization branch now... any idea what is going on with that one?

**Comment by znicolaou** [comment]
I think the revert button will create a new pull request that copies the old one, but yeah, it's a little confusing. I also deleted the branch from the fork. Definitely easier to work with the pull request from the branch here than from the forked branch. 

I'm investigating the pytest failures now--it does seem to be a cvxpy issue. I think they may only appear for certain cvxpy versions.

**Comment by znicolaou** [comment]
I deleted the revert branch-I think I accidentally created that when I tried to see what that revert button does...

**Comment by akaptano** [comment]
Yeah I've seen a few of those cvxpy issues before. Let me handle it... usually this is weirdness from the trapping optimizer so I'll try and fix this with the next few commits.

**Comment by akaptano** [comment]
Okay pushed some changes. I have the weak form mostly working, although there seems to be a bug with the mixed derivative terms and still using more for loops than I would like. It is hard to avoid for loops for some of this because each function needs to get interpolated onto each subdomain before the weak form terms are computed assembled. There also seems to be big with the third order finite differences, when axis is specified. Kind of confused why, but you can see it fail one of the tests by running "pytest" in the test/differentiation/ directory. Getting there!

**Comment by znicolaou** [comment]
Awesome, nice work!  

I also investigated the pytest cvxpy errors a bit, and found that my local environment was using scs package version 2.1.4, but pip chooses to install version 3.0.0 for a new environment. Adding the requirement scs<=2.1.4 to requirements-dev.txt fixes the optimization tests, so I pushed that for now. Now only that last test from the differentiation fails. I'll see if I can figure out why!

**Comment by akaptano** [comment]
Nice! Yeah we have had other issues with SCS, it has a lot of weird dependencies. Sounds good on the fix. Note that the third order differentiation test is comparing differentiation using a uniform timestep dt and corresponding uniform time base (this is to check that the differentiation agrees whether or not you pass a timestep or the whole time series), and it works if the axis is not specified but seems to not work when the axis is specified. But the first and second order tests are working!

**Comment by akaptano** [comment]
Anyways, if we fix that, I'll add a few more differentiation tests with the new features. Would be good to be very sure this all works, since this will cause weak form errors too.

**Comment by znicolaou** [comment]
I found the d=3 problem. The variable timestep uses t_diff=t[4:]-t[:-4] for the instantaneous time step, which needs to rescaled by 1/4 rather than 1/2. I pushed that little change, and the test succeeds!

**Comment by akaptano** [comment]
Yay! Go team. So things left to do:
1. Make sure the weak form mixed derivative/non-derivative terms are calculated correctly and as fast as possible.
2. Update example notebooks 1, 10, and 12 (there is some out of date text now too) and make sure reproducing roughly the old performance or better.
3. Add additional unit tests for all this stuff.
4. Draft a PySINDy release about this awesome fix.

**Comment by akaptano** [comment]
By the way, do you have any preferences about dropping the spectral derivative file? It was unclear to me if there was any difference with your code and the spectral differentiation available through the derivative package.

**Comment by znicolaou** [comment]
Sounds great! I should have some time to check out the weak form and example notebooks in the next day or two. I'm not attached to the spectral derivative at all, but I plan to compare with the differentiation package to see if it performs comparably.

**Comment by akaptano** [comment]
Sounds good :)

By the way, I think I already got at least an order of magnitude speedup (at least a factor of 10 but maybe closer to 50) with the new weak formulation, and I think there are additional speedups left around.

**Comment by znicolaou** [comment]
I checked out your new WeakPDELibrary -- looks awesome! I can tell more or less what's going on, so I'll play a bit with verifying the calculations per your 1) above and seeing if I notice any easy optimizations. But it looks like you did a very good job already :). There are some annoying future warnings related to using tuples rather than lists/arrays for the slice indices, which are easy to fix by casting to tuples. I think there are also a couple updates to the PDELibrary in order, including using the spatiotemporal grid in place of separate space and time grids. 

As per 2), I fixed a couple small errors we introduced in the example notebook 1 -- the threshold in the inequality constrained SR3 example got set to 0 instead of 10, and the differentiation along an axis example overwrote the value of dt, which messed up some of the subsequent fits a bit. Some text still needs updating.

Lastly, I looked into the details of the finite difference derivatives a bit more--I think the variable timestep calculations could be improved a bit. It is using the central differences for a uniform timestep with a window-averaged timestep now, but there is an easy calculation to find the true coefficients for arbitrary stencils, as per https://en.wikipedia.org/wiki/Finite_difference_coefficient. It wouldn't be hard to hand-code higher order derivative orders and accuracy orders also, which I'll look into. It's important to not trade performance for generality here, though, so I'll be careful.

**Comment by akaptano** [comment]
Thanks, yeah getting rid of the FutureWarnings would be great. I actually don't think the PDELibrary object needs the spatiotemporal grid, since it does not need the timebase to compute any of the terms (although we could potentially allow for temporal derivatives in the library, this leads to all the complexity of SINDy-PI so might want to keep that functionality specific to SINDy-PI usage). The WeakPDELibrary terms gets the full spatiotemporal grid because the subdomain integrals are in space and time. 

Thanks for the other random fixes + sounds good about the derivatives if you have the time to investigate. Moving this conversation to the new pull request now!
